PyInstaller/depend/analysis.py

```diff
@@ -156,7 +124,85 @@ def run_script(self, pathname, caller=None):
+        # Check 'module_name' and all its parent packages.
+        parent_packages = [module_name] + module_parent_packages(module_name)
+        for module in parent_packages:
+            # Check if there is any excluded import for the module.
+            if module in self.excluded_imports:
+                # TODO verify if 'imported_by' being subpackage of excludedimports needs checking.
+                # 'imported_by' is a subpackage of 'module', keep the excluded
+                # import and just end the function.
+                #if imported_by.startswith(module):
+                    # Add submodules of excluded modules to the exclude list and
+                    # exclude it for the same modules as the parent package.
+                    #self.excluded_imports[module_name] = self.excluded_imports[module]
+                    #return True
+                # Modules for which the excluded import should be kept.
+                not_allowed_modules = self.excluded_imports[module]
+                # 'imported_by' is not allowed or is subpackage of any not
+                # allowed module.
+                for not_allowed in not_allowed_modules:
+                    if imported_by.startswith(not_allowed):
+                        return True
+                # 'module_name' is excluded for some modules but this time
+                # it is imported by completely different module and thus
+                # 'module_name' and all its parent should be included in the analysis.
+                # Thus remove them from the excluded imports.
+                logger.warn('Including previously excluded %r (required by %s)' % (module_name, imported_by))
+                del self.excluded_imports[module]
+
+        # Module could be imported the usual way.
+        return False
+
+
+    def _process_imports(self, node):
+        """
+        This method wraps the superclass method with support for excludedimports.
+        Wrapping avoids creating MissingModule nodes if any dependency of module
+        'node' imports excluded modules.
+
+        Before importing a module, check first if it should be kept excluded
+        or not. It does not make sense check excluded modules if the caller
+        module is not known.
+
+        Wrapping this method allow to behave like modules from excludedimports
+        were not even recognided as a dependency of a module.
+        """
+        if node._imported_modules is None:
+            return
+
+        new_imported_modules = []
+        imported_by = node.identifier
+
+        for mod_data in node._imported_modules:
+            module_name = mod_data[1][0]
+            # Skip excluded modules and remove it from dependencies.
+            if not self._is_excluded_import(module_name, imported_by):
+                new_imported_modules.append(mod_data)
+        node._imported_modules = new_imported_modules
```

Comment:
@matysek I just had another idea: What about hooking here? Handling excluded import here would be easier, hiddenimports could just be added to the list, datas and binaries do no harm.